Add ramalama gpu_detector

dougsland opened the pull request:
GPUDetector can be used in any part of the code.
However, useful for users to run in the cli.

ramalama --show-gpus-available

## Summary by Sourcery

New Features:
- Users can now list available GPUs and their VRAM using the `--show-gpus-available` CLI flag.

dougsland commented:
cc Brian <bmahabir@bu.edu>

dougsland commented:
in a different patch will take care of macOS env, need to test it but it's very later now 2AM.

rhatdan commented on ramalama/cli.py:
```diff
@@ -189,6 +190,14 @@ def configure_arguments(parser):
         help="offload the workload to the GPU",
     )
+
+    parser.add_argument(
```
This should be part of ramalama info, no option equired, unless this takes a lot of time to call?

dougsland replied:
still want to change the code @rhatdan ?

rhatdan replied:
Yes just add this to ramalama info.

dougsland replied:
Tested with Intel:
```
   "GPUs": {
        "Detected GPUs": [
            {
                "Details": "00:02.0 VGA compatible controller: Intel Corporation TigerLake-H GT1 [UHD Graphics] (rev 01)",
                "Env": "ONEAPI_DEVICE_SELECTOR",
                "GPU": "Intel GPU",
                "VRAM": "Unknown"
            }
        ],
        "Errors": [
            {
                "Error": "Unable to detect NVIDIA GPU(s). Error: Unknown error (check if NVIDIA drivers are loaded).",
                "Vendor": "NVIDIA"
            },
            {
                "Error": "No AMD GPU detected or drivers missing.",
                "Vendor": "AMD"
            }
        ]
    },
```

dougsland replied:
macos:

```
   "GPUs": {
        "Detected GPUs": [
            {
                "Cores": "32",
                "GPU": "Apple M1 Max",
                "Metal": "Metal 3",
                "Vendor": "Apple (0x106b)"
            },
            {
                "GPU": "Color LCD"
            }
        ],
        "Errors": "No errors"
    },
    "Image": "quay.io/ramalama/ramalama",
    "Runtime": "llama.cpp",
    "Store": "/var/lib/ramalama",
    "UseContainer": true,
    "Version": "0.5.4"
}

```

dougsland replied:
@rhatdan I think we addressed this. AFAIK, macOS with Apple M(family) cannot have Intel or AMD together. However, it's possible to have hardware with AMD and Intel (not in the mainboard) as extra card. 
Please let me know your thoughts. I thought this change would be super easy but in the end, not :)

rhatdan commented on ramalama/gpu_detector.py:
```diff
@@ -0,0 +1,117 @@
+                except ValueError as ex:
+                    print(f"Error parsing Nvidia GPU info: {ex}")
+        except FileNotFoundError:
+            print("Warning: `nvidia-smi` not found. No NVIDIA GPU detected or drivers missing.")
```
Pleas raise an exception here. And handle this at a higher level.  Warnings should be printed to stderr.

rhatdan commented on ramalama/gpu_detector.py:
```diff
@@ -0,0 +1,117 @@
+            print("Warning: `nvidia-smi` not found. No NVIDIA GPU detected or drivers missing.")
+        except subprocess.CalledProcessError as e:
+            error_msg = e.stderr.strip() if e.stderr else "Unknown error (check if NVIDIA drivers are loaded)."
+            print(f"Unable to detect NVIDIA GPU(s). Error: {error_msg}")
```
Raise error.

rhatdan commented on ramalama/gpu_detector.py:
```diff
@@ -0,0 +1,117 @@
+                    self.best_env = "ONEAPI_DEVICE_SELECTOR"
+
+        except FileNotFoundError:
+            print("Warning: `lspci` not found. Unable to detect Intel GPU.")
```
Raise error.

dougsland commented:
CI/CD error seems not related to the PR.

dougsland commented:
I don't have permission to re-run CI/CD. Someone can re-run please?  @ericcurtin

rhatdan commented on ramalama/cli.py:
```diff
@@ -238,9 +248,32 @@ def configure_subcommands(parser):
+            return
+        print("No compatible GPUs found.")
+    except RuntimeError as e:
+        print(f"Error: {e}", file=sys.stderr)  # Gracefully print the error
```
Do not catch the error allow bin/ramalama to handle it.

rhatdan commented on ramalama/gpu_detector.py:
```diff
@@ -0,0 +1,168 @@
+            raise RuntimeError(f"Unable to detect macOS GPU(s). Error: {error_msg}")
+
+    def detect_best_gpu(self, gpu_template):
+        """Compares Nvidia, AMD, and Intel GPUs and appends the best GPU with the highest VRAM to gpu_template.
```
What about MAC gpu?

dougsland replied:
It works, let me update. 

Linux:
```
ramalama --show-gpus-available 
Warning: Unable to detect NVIDIA GPU(s). Error: Unknown error (check if NVIDIA drivers are loaded).
GPU(s) detected:
  - Index: Intel GPU (unknown memory), VRAM: 0 MiB, Env: ONEAPI_DEVICE_SELECTOR
```

macOS:
```
ramalama --show-gpus-available
GPU(s) detected:
  - Index: Apple M1 Max, VRAM: 8192 MiB, Env: METAL_DEVICE_INDEX
```

dougsland commented:
Thinking here I believe maybe instead of ERROR we should print INFO to avoid confusion. When found one GPU and try to get the others but don't find.  @rhatdan @ericcurtin

dougsland commented:
> Thinking here I believe maybe instead of ERROR we should print INFO to avoid confusion. When found one GPU and try to > get the others but don't find. @rhatdan @ericcurtin. At the same time, it's getting more and more code, I would merge this
> one and guarantee we have something tested fine and a next patch I change to INFO if all agree with the idea.

Already changed to INFO, should be all set.

dougsland commented:
Intel that I have:

```
  "GPUs": {
        "Detected GPUs": [
            {
                "Details": "00:02.0 VGA compatible controller: Intel Corporation TigerLake-H GT1 [UHD Graphics] (rev 01)",
                "Env": "ONEAPI_DEVICE_SELECTOR",
                "GPU": "Intel GPU",
                "VRAM": "Unknown"
            }
        ],
        "INFO": [
            {
                "INFO": "Unable to detect NVIDIA GPU(s). Error: Unknown error (check if NVIDIA drivers are loaded).",
                "Vendor": "NVIDIA"
            },
            {
                "INFO": "No AMD GPU detected or drivers missing.",
                "Vendor": "AMD"
            }
        ]
    },
    "Image": "quay.io/ramalama/ramalama",
    "Runtime": "llama.cpp",
    "Store": "/home/douglas/.local/share/ramalama",
    "UseContainer": true,
    "Version": "0.5.4"
}
```

macOS:

```
 "GPUs": {
        "Detected GPUs": [
            {
                "Cores": "32",
                "GPU": "Apple M1 Max",
                "Metal": "Metal 3",
                "Vendor": "Apple (0x106b)"
            },
            {
                "GPU": "Color LCD"
            }
        ],
        "INFO": "No errors"
    },
    "Image": "quay.io/ramalama/ramalama",
    "Runtime": "llama.cpp",
    "Store": "/Users/douglaslandgraf/.local/share/ramalama",
    "UseContainer": false,
    "Version": "0.5.4"
```

maxamillion commented:
This looks great. Thank you for the contribution and for addressing all the feedback. :smile:

maxamillion commented on ramalama/cli.py:
```diff
@@ -238,6 +240,58 @@ def configure_subcommands(parser):
     version_parser(subparsers)
 
 
+def show_gpus_available_cli(args):
```
unless I'm mistaken, this appears to not be registered via `parser.add_argument`

dougsland replied:
@maxamillion we don't use --show-gpus-available anymore (@rhatdan requested, if request for me is priority ;-) ). We are now only using `ramalama info`

dougsland commented:
> This looks great. Thank you for the contribution and for addressing all the feedback. 😄

My pleasure, looking forward to contribute more. Please share issues that you guys looking for help.

rhatdan commented:
Great work @dougsland 

LGTM
